### ISSUE TITLE
`st_read` calls `as.data.frame` which changes column names without warning

### ISSUE THREAD

**Issue by rgzn** [r-spatial/sf#1916]
When reading external data using `st_read`, field names in data sources are changed by `as.data.frame`. When converting field names to sfc column names, it removes special characters such as "-" and "(", replacing them with ".". This is especially problematic because:

- no warning is given
- this naming convention is not enforced in any of the ogr compatible data sources
- `st_write` does not do this, making it non-commutable with `st_read`

This happens here:

https://github.com/r-spatial/sf/blob/c8d868c39ad197201ab0e172b1a6f2b0d1cdd347/R/read.R#L179

and could be fixed by adding the following parameter to the call to `as.data.frame`:
`optional = TRUE`

Alternatively, this could be a parameter in `st_read` for the user to set. 

Here is an example to test:

```
library(dplyr)
library(sf)

st_read(system.file("shape/nc.shp", package="sf")) %>% 
  rename_with(~ gsub("_","-",.x, fixed= TRUE)) ->
  nc_dash

nc_dash %>%
  st_write("nc_dashes.geojson")

st_read("nc_dashes.geojson") -> 
  nc_dash_read

names(nc_dash) == names(nc_dash_read)
```

**Comment by edzer** [comment]
We now see:
```r
st_read("nc_dashes.geojson", optional = TRUE) -> 
  nc_dash_read
# Reading layer `nc_dashes' from data source `/tmp/nc_dashes.geojson' using driver `GeoJSON'
# Simple feature collection with 100 features and 14 fields
# Geometry type: MULTIPOLYGON
# Dimension:     XY
# Bounding box:  xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
# Geodetic CRS:  NAD27

all(names(nc_dash) == names(nc_dash_read))
# [1] TRUE
```
so this is now possible. Making `optional` default to `TRUE` would require a full reverse dependency check, and communication with possibly breaking dependent packages, but is also up for discussion, as it is against how `data.frame`s are created by default. Having it `TRUE` when creating a `tibble` might make sense, I assume `tibble`s don't make "valid" names?

**Comment by rsbivand** [comment]
It would be much better to call any new argument check.names, not optional, as in data.table() rather than the coercion method. If the user wants tibble-type behaviour, use read_sf(), not st_read().

**Comment by rgzn** [comment]
Thanks Edzer!

Yeah, this seems good. I would maybe suggest some sort of warning message being thrown if names are changed as well, though because `as.data.frame` doesn't give you one, this might be hard to do. It's a little bit confusing when you pull in GIS data from an external source, process it, write the output, and then it's no longer compatible because the field names have been changed. 

@rsbivand 
> It would be much better to call any new argument check.names, not optional, as in data.table() rather than the coercion method. If the user wants tibble-type behaviour, use read_sf(), not st_read().

Shouldn't `st_read(as_tibble = TRUE, ...)` also deliver tibble-typen behavior? I agree the `optional` argument inherited from `as.data.frame` is not a very informative name.

**Comment by rsbivand** [comment]
No tibbles should be returned by any standard function, they mascarade as data.frame objects but are not such. Keep them in read_sf() or coerce. No warning is required for making column names valid. Trapping the GDAL object field names is possible, so a message might be feasible, but adds to read time.

**Comment by rgzn** [comment]
I see. So you are saying the `as_tibble` parameter should not be an option?

**Comment by rsbivand** [comment]
Exactly, or at least modify `read_sf() ` to coerce to `"sf-tibble"` for those who may want that structure, rather than use an argument. The former is supposed to appear tidy, so that is where tibbles might be expected, as its help page entry asserts. I always coerce any tibbles created by code people send me to data.frame objects to avoid confusion later when doing statistical analysis.

**Comment by rsbivand** [comment]
That suggests that `optional=` be changed to `check.names=` default TRUE in `st_read()`, and that `read_sf()` calls `st_read()` with `check.names=FALSE` and `as_tibble=TRUE`, so `st_read()` behaves normally and `read_sf()` always treats the returned object as an sf-tibble:

```
tf <- tempfile(fileext=".csv")
library(readr)
write_csv(nc_dash, tf)
read_csv(tf)
```

**Comment by edzer** [comment]
See https://github.com/r-spatial/sf/commit/1558d2ee5ca86ef8abc53a113db726594dae3f91

@roger the lines you point to happen when there is _no_ geometry column, so quite rarely. I now settled on passing `optional || as_tibble` as value to the `optional` argument in the "right" `as.data.frame` call, line 180, which then also gives
```r
read_sf("nc_dashes.geojson") -> nc_dash_read

all(names(nc_dash) == names(nc_dash_read))
# [1] TRUE
```
I see no need to pass `.name_repair` - people with understanding at that level know how to postprocess the object read.

**Comment by rsbivand** [comment]
Should I run revdeps?

**Comment by edzer** [comment]
Let's do that shortly before 1.0-8 gets released, 1.0-7 just appeared on CRAN.
